pkg/storage/gcs.go

```diff
@@ -0,0 +1,140 @@
+	gcs.StorageClass = options.StorageClass
+	gcs.PredefinedAcl = options.PredefinedAcl
+
+	b, err := ioutil.ReadFile(options.CredentialsFile)
+	if err != nil {
+		return err
+	}
+	gcs.CredentialsBlob = string(b)
+	return nil
+}
+
+func defineGCSFlags(flags *pflag.FlagSet) {
+	flags.String(gcsEndpointOption, "", "Set the GCS endpoint URL")
+	flags.String(gcsStorageClassOption, "", "Specify the GCS storage class for objects. If it is not set, objects uploaded are followed by the default storage class of the bucket. See https://cloud.google.com/storage/docs/storage-classes for valid values.")
+	flags.String(gcsPredefinedAcl, "", "Specify the GCS predefined acl for objects. If it is not set, objects uploaded are followed by the acl of bucket scope. See https://cloud.google.com/storage/docs/access-control/lists#predefined-acl for valid values.")
+	flags.String(gcsCredentialsFile, "", "Set the GCS credentials file path. You can get one from https://console.cloud.google.com/apis/credentials.")
+
+	_ = flags.MarkHidden(gcsEndpointOption)
+}
+
+func getBackendOptionsFromGCSFlags(flags *pflag.FlagSet) (options GCSBackendOptions, err error) {
+	options.Endpoint, err = flags.GetString(s3EndpointOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	options.StorageClass, err = flags.GetString(gcsStorageClassOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	options.PredefinedAcl, err = flags.GetString(gcsPredefinedAcl)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	options.CredentialsFile, err = flags.GetString(gcsCredentialsFile)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	return
+}
+
+type gcsStorage struct {
+	gcs    *backup.GCS
+	bucket *storage.BucketHandle
+}
+
+// Write file to storage
+func (s *gcsStorage) Write(ctx context.Context, name string, data []byte) error {
+	object := s.gcs.Prefix + name
+	wc := s.bucket.Object(object).NewWriter(ctx)
+	wc.StorageClass = s.gcs.StorageClass
+	wc.PredefinedACL = s.gcs.PredefinedAcl
+	_, err := wc.Write(data)
+	if err != nil {
+		return err
+	}
+	return wc.Close()
+}
+
+// Read storage file
+func (s *gcsStorage) Read(ctx context.Context, name string) ([]byte, error) {
+	object := s.gcs.Prefix + name
+	rc, err := s.bucket.Object(object).NewReader(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer rc.Close()
+
+	b := make([]byte, rc.Attrs.Size)
+	_, err = io.ReadFull(rc, b)
+	return b, err
+}
+
+// FileExists return true if file exists
+func (s *gcsStorage) FileExists(ctx context.Context, name string) (bool, error) {
+	object := s.gcs.Prefix + name
+	_, err := s.bucket.Object(object).Attrs(ctx)
+	if err != nil {
+		if err == storage.ErrObjectNotExist {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func newGCSStorage(gcs *backup.GCS) (*gcsStorage, error) {
+	var clientOps []option.ClientOption
+	clientOps = append(clientOps, option.WithCredentialsJSON([]byte(gcs.GetCredentialsBlob())))
```

Comment:
It looks like we are being forced into one particular way of doing credentials here. I am actually not sure what this is doing. But on GCE or GKE with the workload identity feature enabled the SDK will already automatically find credentials in the metadata.

Reply:
Since `br` is not sure that all the TiKV instances are deployed on GCE or GKE either, the simplest way is getting a credentials JSON and sending it to other TiKV instances.

Yes, I think it's a cool thing to use service account from environment on GCE or GKE. Maybe in the future and we have a idea to handle the credentials for TiKV instances.

Reply:
I've created an issue https://github.com/pingcap/br/issues/106

Reply:
It happens automatically if you don't force one particular method like this. So you don't have to do anything to support it (other than not hard-code a particular auth technique).

Reply:
If BR is not deployed on google cloud but on someone's IDC, we need this credentials file.

Reply:
Right, so I think all you need here is an if statement, something like:

``` go
creds := []byte(gcs.GetCredentialsBlob()
if creds != "" {
     	clientOps = append(clientOps, option.WithCredentialsJSON(creds))
}
```

Reply:
I'll handle this on #106 because we need to deal with `--send-credentials-to-tikv` option here.